Implement more of InGameUI

jonwil opened the pull request.

jonwil committed:
Implement more of InGameUI

xezon commented on src/hooker/setuphooks_zh.cpp:
```diff
@@ -3771,6 +3773,13 @@ void Setup_Hooks()
+    Hook_Any(0x007A95B0, W3DInGameUI::Update);
+    Hook_Any(0x007A95D0, W3DInGameUI::Draw);
+    Hook_Any(0x007A9650, W3DInGameUI::Draw_Selection_Region);
+    // Hook_Any(0x007A9690, W3DInGameUI::Draw_Move_Hints);
```
Is this commented on purpose?

jonwil replied:
Fixed.

xezon commented on src/hooker/setuphooks_zh.cpp:
```diff
@@ -1,4 +1,5 @@
 ﻿/**
+/**
```
Can remove.

jonwil replied:
fixed.

xezon commented on src/platform/w3dengine/client/w3dingameui.cpp:
```diff
@@ -13,9 +13,225 @@
+
+void W3DInGameUI::Draw_Selection_Region()
+{
+    g_theDisplay->Draw_Open_Rect(m_dragSelectRegion.lo.x,
```
```c++
virtual void Draw_Open_Rect(int, int, int, int, float, unsigned) = 0;
```
to
```c++
virtual void Draw_Open_Rect(int x, int y, int w, int h, float line, unsigned int color) = 0;
```

Or use the names from

```c++
void W3DDisplay::Draw_Open_Rect(int32_t x, int32_t y, int32_t width, int32_t height, float border_width, uint32_t color)
```

xezon commented on src/platform/w3dengine/client/w3dingameui.cpp:
```diff
@@ -13,9 +13,225 @@
+                W3DDisplay::s_3DScene->Remove_Render_Object(m_moveHintRenderObj[count]);
+            }
+
+            goto label1;
```
Can remove `goto` by writing here

```c++
if (++count >= MAX_MOVE_HINTS) {
    return;
}
continue;
```

jonwil replied:
fixed.

xezon commented on src/platform/w3dengine/client/w3dingameui.cpp:
```diff
@@ -13,9 +13,225 @@
+void W3DInGameUI::Draw_Place_Angle(View *view)
+{
+    if (m_locator01 == nullptr) {
+        m_locator01 = W3DDisplay::s_assetManager->Create_Render_Obj("Locater01");
```
Not `"Locator01"` ?

jonwil replied:
Locater is correct spelling here, I renamed some variables.

xezon commented on src/platform/w3dengine/client/w3dingameui.cpp:
```diff
@@ -13,9 +13,225 @@
+    }
+
+    if (m_locator02 == nullptr) {
+        m_locator02 = W3DDisplay::s_assetManager->Create_Render_Obj("Locater02");
```
Not `"Locator02"` ?

jonwil replied:
Locater is correct spelling here, I renamed some variables.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -255,6 +258,161 @@ void Toggle_Replay_Controls()
     }
 }
 
+InGameUI::InGameUI() :
```
`m_dragSelectRegion`
`m_scrollAmt`
`m_isQuitMenuVisible`
`m_cameraRotateLeft`
`m_cameraRotateRight`
`m_cameraZoomIn`
`m_cameraDrawingTrackable`
`m_cameraZoomOut`
`m_noRadarEdgeSound`

are not initialized

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -255,6 +258,161 @@ void Toggle_Replay_Controls()
+        m_buildProgress[i].m_control = nullptr;
+    }
+
+    m_placeIcon = new Drawable *[g_theWriteableGlobalData->m_maxLineBuildObjects];
```
Can be optimized to

```c++
m_placeIcon = new Drawable *[g_theWriteableGlobalData->m_maxLineBuildObjects]();
```

for 0 initialization.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -255,6 +258,161 @@ void Toggle_Replay_Controls()
+    m_namedTimerNormalBold(false),
+    m_namedTimerReadyFont("Arial"),
+    m_namedTimerReadyPointSize(10),
+    m_namedTimerReadyBold(0),
```
false

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -255,6 +258,161 @@ void Toggle_Replay_Controls()
+    m_drawableCaptionFont("Arial"),
+    m_drawableCaptionPointSize(10),
+    m_drawableCaptionBold(false),
+    m_tooltipsDisabled(false),
```
0 ?
Type is put as unsigned int.

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -255,6 +258,161 @@ void Toggle_Replay_Controls()
+    m_placeAnchorEnd.y = 0;
+
+    for (int i = 0; i < MAX_UI_MESSAGES; i++) {
+        m_uiMessages[i].full_text.Clear();
```
Is obsolete call.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+        m_currentlyPlayingMovie = movie_name;
+        m_cameoVideoBuffer = g_theDisplay->Create_VideoBuffer();
+
+        if (m_cameoVideoBuffer == nullptr
```
!= nullptr

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+    GameWindow *hud =
+        g_theWindowManager->Win_Get_Window_From_Id(nullptr, g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:RightHUD"));
+    hud->Win_Get_Instance_Data()->Set_VideoBuffer(nullptr);
+    if (m_cameoVideoBuffer != nullptr) {
```
I suggest add blank line above this.

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+    if (ret == -1) {
+        g_theInGameUI->Message(g_theGameText->Fetch("GUI:NothingSelected"));
+    } else if (ret != 0) {
+        g_theInGameUI->Message(g_theGameText->Fetch("GUI:SelectedAcrossScreen"));
```
GUI:SelectedAcrossMap ?

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+        Drawable *draw = g_theInGameUI->Get_First_Selected_Drawable();
+
+        if (draw == nullptr || draw->Get_Object() != nullptr || !draw->Get_Object()->Is_KindOf(KINDOF_STRUCTURE)) {
+            g_theInGameUI->Message(g_theGameText->Fetch("GUI:SelectedAcrossScreen"));
```
GUI:SelectedAcrossMap ?

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+            return false;
+        }
+
+        bool equivalent = obj->Get_Template()->Is_Equivalent_To(tmplate);
```
is_equivalent

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+    if (ret == -1) {
+        g_theInGameUI->Message(g_theGameText->Fetch("GUI:NothingSelected"));
+    } else if (ret != 0) {
+        g_theInGameUI->Message(g_theGameText->Fetch("GUI:SelectedAcrossScreen"));
```
GUI:SelectedAcrossMap ?

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+        Drawable *draw = g_theInGameUI->Get_First_Selected_Drawable();
+
+        if (draw == nullptr || draw->Get_Object() != nullptr || !draw->Get_Object()->Is_KindOf(KINDOF_STRUCTURE)) {
+            g_theInGameUI->Message(g_theGameText->Fetch("GUI:SelectedAcrossScreen"));
```
GUI:SelectedAcrossMap ?

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+            } else {
+                for (Drawable *drawable = g_theGameClient->First_Drawable(); drawable != nullptr;
+                     drawable = drawable->Get_Next()) {
+                    count += Similar_Unit_Selection(drawable, &data);
```
This is adding bool value to count.

Perhaps meant to be

```c++
if (Similar_Unit_Selection(drawable, &data))
    count++;
```
?

jonwil replied:
Nope, original code does this.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+{
+    KindOfUnitSelectionData data;
+    int count = 0;
+    int size = Get_All_Selected_Drawables()->size();
```
previous_count

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+FloatingTextData::FloatingTextData() : m_color(0), m_frameTimeOut(0), m_frameCount(0)
+{
+    m_pos3D.Zero();
+    m_text.Clear();
```
Is obsolete call.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+    }
+}
+
+void InGameUI::Disable_Tooltips_Until(unsigned int until)
```
frame

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+        m_idleWorkerWin = g_theWindowManager->Win_Get_Window_From_Id(
+            nullptr, g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:ButtonIdleWorker"));
+        captainslog_dbgassert(
+            m_idleWorkerWin != nullptr, "InGameUI::Show_Idle_Worker_Layout could not find IdleWorker.wnd to load ");
```
Looks like `m_idleWorkerCount = Get_Idle_Worker_Count();` is missing in this branch.

jonwil replied:
Original doesn't have it.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+    return nullptr;
+}
+
+void InGameUI::Show_Idle_Worker_Layout()
```
No call to `Gadget_Button_Set_Text` in this function?

jonwil replied:
Original doesn't have it.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+{
+    for (auto it = m_windowLayoutList.begin(); it != m_windowLayoutList.end(); it++) {
+        if (*it == layout) {
+            m_windowLayoutList.erase(it);
```
`return`, `break` or `it =` missing.

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+bool InGameUI::Can_Selected_Objects_Non_Attack_Interact_With_Object(
+    const Object *object_to_interact_with, SelectionRules rule)
+{
+    for (int i = ACTIONTYPE_ATTACK_OBJECT; i < NUM_ACTIONTYPES; i++) {
```
This looks odd. Effectively this loop is equivalent to simplified:

```c++
    for (int i = ACTIONTYPE_ATTACK_OBJECT + 1; i < NUM_ACTIONTYPES; i++) {
        if (Can_Selected_Objects_Do_Action(static_cast<ActionType>(i), object_to_interact_with, rule, false)) {
            return true;
        }
    }
```

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+    }
+
+    const std::list<Drawable *> *list = g_theInGameUI->Get_All_Selected_Drawables();
+    int count = 0;
```
```

ATTACK_RESULT_0 = CANNOT_ATTACK
ATTACK_RESULT_1 = CAN_ATTACK_BUT_UNREACHABLE
ATTACK_RESULT_2 = CAN_ATTACK_BUT_OUT_OF_RANGE
ATTACK_RESULT_3 = CAN_ATTACK
```

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+
+    const std::list<Drawable *> *list = g_theInGameUI->Get_All_Selected_Drawables();
+    int count = 0;
+    CanAttackResult result1 = ATTACK_RESULT_0;
```
any_can_attack

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+    const std::list<Drawable *> *list = g_theInGameUI->Get_All_Selected_Drawables();
+    int count = 0;
+    CanAttackResult result1 = ATTACK_RESULT_0;
+    CanAttackResult result2 = ATTACK_RESULT_3;
```
all_can_attack

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+            }
+            case ACTIONTYPE_COMBAT_DROP:
+                can_do = g_theActionManager->Can_Enter_Object(
+                    draw->Get_Object(), object_to_interact_with, COMMANDSOURCE_PLAYER, CAN_ENTER_2);
```
CAN_ENTER_2 = CAN_ENTER_WITH_COMBAT_DROP

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+    Coord3D *position,
+    SelectionRules rule,
+    unsigned int options,
+    Object *o) const
```
custom_object

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+        draw = nullptr;
+    }
+
+    std::list<Drawable *> list;
```
custom_drawables

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+        list.push_back(draw);
+    }
+
+    const std::list<Drawable *> *list2;
```
drawables

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+        return false;
+    }
+
+    Drawable *draw;
```
custom_drawable

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+        Drawable *drawable = *it;
+        count++;
+
+        if (needs_target_object || needs_target_pos) {
```
Overcomplicated condition tree.

Simpler:

```c++
if (needs_target_object) {
...
} else if (needs_target_pos) {
...
} else {
...
}
```

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+                }
+            }
+        } else {
+            if (g_theActionManager->Can_Do_Special_Power(draw->Get_Object(), power, COMMANDSOURCE_PLAYER, options, 1)) {
```
`drawable`, no ?
Otherwise this may be null.

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+        return false;
+    }
+
+    const std::list<Drawable *> *list = g_theInGameUI->Get_All_Selected_Drawables();
```
drawables

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+        return ATTACK_RESULT_0;
+    }
+
+    const std::list<Drawable *> *list = g_theInGameUI->Get_All_Selected_Drawables();
```
drawables

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+{
+    int count = 0;
+    int qualify = 0;
+    const std::list<Drawable *> *list = g_theInGameUI->Get_All_Selected_Drawables();
```
drawables

jonwil replied:
fixed.

xezon commented on src/game/client/ingameui.cpp:
```diff
@@ -2993,3 +3151,1015 @@ void InGameUI::Update_And_Draw_World_Animations()
+
+    for (auto it = list->begin(); it != list->end(); it++) {
+        Drawable *draw = *it;
+        if (needs_target_object || needs_target_pos) {
```
Overcomplicated condition tree.

Simpler:

```c++
if (needs_target_object) {
...
} else if (needs_target_pos) {
...
} else {
...
}
```

xezon commented on src/platform/w3dengine/client/w3dingameui.cpp:
```diff
@@ -13,9 +13,225 @@
+            break;
+        }
+
+    label2:
```
Is there no version of this function without this label?

jonwil replied:
I can't find anything cleaner.

xezon reviewed (commented):
Game crashes when pressing Idle Worker button:

```
>	thyme.dll!InGameUI::Select_Next_Idle_Worker() Line 3646	C++
 	thyme.dll!GameWindowManager::Win_Send_System_Msg(GameWindow * window, unsigned int msg, unsigned int data1, unsigned int data2) Line 451	C++
 	game.dat!005abdb2()	Unknown
 	game.dat![Frames below may be incorrect and/or missing, no symbols loaded for game.dat]	Unknown
 	thyme.dll!GameWindowManager::Win_Send_Input_Msg(GameWindow * window, unsigned int msg, unsigned int data1, unsigned int data2) Line 464	C++
 	thyme.dll!GameWindowManager::Win_Process_Mouse_Event(GameWindowMessage msg, ICoord2D * mouse_pos, void * data) Line 742	C++
 	game.dat!0061725f()	Unknown
 	thyme.dll!MessageStream::Propagate_Messages() Line 177	C++
```

```
Exception thrown: read access violation.
draw was nullptr.
```

```c++
            Drawable *draw = g_theInGameUI->Get_First_Selected_Drawable();

            for (auto it = m_idleWorkerLists[index].begin(); it != m_idleWorkerLists[index].end(); it++) {
                if (*it == draw->Get_Object()) {
                    if (it != m_idleWorkerLists[index].end()) {
                        obj = *it;
                    } else {
                        obj = *m_idleWorkerLists[index].begin();
                    }

                    break;
                }
            }
```

xezon commented:
Perhaps forgot to hook constructor of InGameUI?

xezon commented:
Selecting faction building and pressing E button will attempt selecting matching units. Is introduced with this change. Is not fixed by commenting all new hooks.

![shot_20240201_182931_1](https://github.com/TheAssemblyArmada/Thyme/assets/4720891/4e58ec1a-12d6-4945-b534-08d8a2574d09)

xezon commented:
USA Sentry Drone + USA Humvee cannot attack together.

Issue appears to be introduced by `InGameUI::Get_Can_Selected_Objects_Attack`.

EDIT: Also happens in Original. Is related to order of construction. If Sentry Drone is constructed first, then Humvee cannot attack in group selection.

![shot_20240201_191422_1](https://github.com/TheAssemblyArmada/Thyme/assets/4720891/0c77dc98-0365-4095-9b17-80f0d9ce1a51)

jonwil commented:
Fixed idle worker crash.
Fixed issue with selecting a structure and pressing E.
InGameUI constructor is hooked in via W3DGameClient::Create_InGameUI

xezon reviewed (commented):
Select next idle worker does not work. It will only select the first worker on every button press.

![shot_20240201_195707_1](https://github.com/TheAssemblyArmada/Thyme/assets/4720891/bbbceb7c-6889-48aa-b89c-c4bd393172ec)

jonwil commented:
Fixed idle worker selection.

xezon reviewed (approved):
Select matching units on screen and map works
Region select and draw works
Floating text works
Draw move hint works
Draw place angle works
Group repair works
Group docking works
Group healing works
Group resume construction works
Group hijack works
Group enter works
Group enter bomb car works
Group disable vehicle works
Set Rally point works
Group combat drop works
Group sabotage building works
Special Power at target works
Special Power at location works
Group attack works
Hidden tool tip during military briefing works
Select next and previous idle worker works
Hide worker idle button in Replay works
